// eslint-plugin-style/dist/util/ast.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.onlyWhiteSpaceBetween = exports.linesBetween = exports.extrema = exports.isTypeImportOrExport = exports.exportModules = exports.importModules = void 0;
function importModules(source) {
    return source.ast.body.filter((node) => node.type === 'ImportDeclaration');
}
exports.importModules = importModules;
function exportModules(source) {
    return source.ast.body.filter((node) => (node.type === 'ExportNamedDeclaration' || node.type === 'ExportAllDeclaration') && Boolean(node.source));
}
exports.exportModules = exportModules;
function isTypeImportOrExport(node) {
    return (Reflect.has(node, 'importKind') && node.importKind === 'type') || (Reflect.has(node, 'exportKind') && node.exportKind === 'type');
}
exports.isTypeImportOrExport = isTypeImportOrExport;
function extrema(source) {
    return [source[0], source[source.length - 1]];
}
exports.extrema = extrema;
function linesBetween(a, b) {
    if (!a.loc || !b.loc) {
        throw new Error(`error: AST was generated without node location information`);
    }
    return b.loc.start.line - a.loc.end.line - 1;
}
exports.linesBetween = linesBetween;
function onlyWhiteSpaceBetween(a, b, source) {
    return !source.getFirstTokenBetween(a, b, { includeComments: true });
}
exports.onlyWhiteSpaceBetween = onlyWhiteSpaceBetween;
