// eslint-plugin-style/dist/util/sort.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sortByPath = void 0;
function isIndexable(value) {
    return typeof value === 'object' && value !== null;
}
function readPath(from, path) {
    if (!isIndexable(from)) {
        return;
    }
    if (path.length === 1) {
        return from[path[0]];
    }
    const [head, ...rest] = path;
    return readPath(from[head], rest);
}
function sort(options) {
    return (a, b) => {
        if (typeof a.sortValue === 'number' && typeof b.sortValue === 'number') {
            return a.sortValue - b.sortValue;
        }
        if (typeof a.sortValue === 'string' && typeof b.sortValue === 'string') {
            return a.sortValue.localeCompare(b.sortValue, options === null || options === void 0 ? void 0 : options.locales, options);
        }
        return 0;
    };
}
// eslint-disable-next-line @typescript-eslint/ban-types
function sortByPath(sources, path, options) {
    const caseGroups = sources.reduce((groups, source) => {
        const sortValue = readPath(source, path);
        const sortable = {
            source,
            sortValue,
        };
        let target = groups.lower;
        if ((options === null || options === void 0 ? void 0 : options.caseGroups) && typeof sortValue === 'string' && sortValue.length) {
            if (/^[@$_]/u.test(sortValue)) {
                target = groups.punctuation;
            }
            else if (sortValue[0].toLocaleUpperCase(options.locales) === sortValue[0]) {
                target = groups.upper;
            }
        }
        target.push(sortable);
        return groups;
    }, {
        punctuation: [],
        upper: [],
        lower: [],
    });
    caseGroups.punctuation.sort(sort(options));
    caseGroups.upper.sort(sort(options));
    caseGroups.lower.sort(sort(options));
    return [
        ...caseGroups.punctuation.map(({ source }) => source),
        ...((options === null || options === void 0 ? void 0 : options.caseFirst) === 'upper' ? caseGroups.upper : caseGroups.lower).map(({ source }) => source),
        ...((options === null || options === void 0 ? void 0 : options.caseFirst) === 'upper' ? caseGroups.lower : caseGroups.upper).map(({ source }) => source),
    ];
}
exports.sortByPath = sortByPath;
