// eslint-plugin-style/dist/rules/group-imports.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = exports.TypeImportConfiguration = exports.ModuleClass = void 0;
const tslib_1 = require("tslib");
const path_1 = require("path");
const builtin_modules_1 = (0, tslib_1.__importDefault)(require("builtin-modules"));
const ast_1 = require("../util/ast");
const rule_1 = require("../util/rule");
const sort_1 = require("../util/sort");
var ModuleClass;
(function (ModuleClass) {
    ModuleClass["Node"] = "node";
    ModuleClass["External"] = "external";
    ModuleClass["Absolute"] = "absolute";
    ModuleClass["Relative"] = "relative";
})(ModuleClass = exports.ModuleClass || (exports.ModuleClass = {}));
var TypeImportConfiguration;
(function (TypeImportConfiguration) {
    TypeImportConfiguration["Include"] = "include";
    TypeImportConfiguration["Exclude"] = "exclude";
    TypeImportConfiguration["Only"] = "only";
})(TypeImportConfiguration = exports.TypeImportConfiguration || (exports.TypeImportConfiguration = {}));
const defaultConfiguration = [
    { class: ModuleClass.Node },
    { class: ModuleClass.External },
    { class: ModuleClass.Absolute },
    { class: ModuleClass.Relative },
];
function groupIndex(node, groups) {
    if (typeof node.source.value !== 'string') {
        return groups.length;
    }
    const importPath = node.source.value;
    const findIndex = (callback) => groups.findIndex((group) => {
        if (group instanceof Array) {
            return group.find((configuration) => callback(configuration));
        }
        return callback(group);
    });
    const isTypeImport = (0, ast_1.isTypeImportOrExport)(node);
    const hardCodedIndex = findIndex((group) => {
        if (typeof group === 'string') {
            return importPath.startsWith(group);
        }
        if (!Reflect.has(group, 'path')) {
            return false;
        }
        if (isTypeImport) {
            return importPath.startsWith(group.path) && group.types !== TypeImportConfiguration.Exclude;
        }
        return importPath.startsWith(group.path) && group.types !== TypeImportConfiguration.Only;
    });
    if (hardCodedIndex >= 0) {
        return hardCodedIndex;
    }
    const pathSegments = importPath.split('/');
    const module = importPath.startsWith('/') ? `/${pathSegments[1]}` : pathSegments[0];
    let moduleClass = ModuleClass.External;
    if (builtin_modules_1.default.includes(module.replace(/^node:/u, ''))) {
        moduleClass = ModuleClass.Node;
    }
    else if (/^(?:\/|\.)/u.exec(module)) {
        moduleClass = (0, path_1.isAbsolute)(module) ? ModuleClass.Absolute : ModuleClass.Relative;
    }
    const classIndex = findIndex((group) => {
        if (typeof group !== 'object' || !Reflect.has(group, 'class')) {
            return false;
        }
        if (isTypeImport) {
            return group.class === moduleClass && group.types !== TypeImportConfiguration.Exclude;
        }
        return group.class === moduleClass && group.types !== TypeImportConfiguration.Only;
    });
    return classIndex >= 0 ? classIndex : groups.length;
}
function checkLines(context, previous, next, lineCount) {
    if ((0, ast_1.linesBetween)(previous, next) === lineCount) {
        return;
    }
    context.report({
        node: previous,
        message: `Expected ${lineCount} empty line${lineCount === 1 ? '' : 's'} after import`,
        fix(fixer) {
            if (!previous.range || !next.range || !(0, ast_1.onlyWhiteSpaceBetween)(previous, next, context.getSourceCode())) {
                return null;
            }
            return fixer.replaceTextRange([previous.range[1], next.range[0]], ''.padEnd(lineCount + 1, '\n'));
        },
    });
}
function groupLabels(groups) {
    return groups.map((group) => {
        if (group instanceof Array || typeof group === 'string' || Reflect.has(group, 'path')) {
            return 'custom';
        }
        return group.class.toUpperCase();
    });
}
exports.rule = {
    meta: {
        fixable: 'code',
        schema: {
            definitions: {
                typeImportConfiguration: {
                    enum: ['include', 'exclude', 'only'],
                },
                moduleConfiguration: {
                    oneOf: [
                        { type: 'string' },
                        {
                            type: 'object',
                            properties: {
                                class: {
                                    enum: ['node', 'external', 'absolute', 'relative'],
                                },
                                types: {
                                    $ref: '#/definitions/typeImportConfiguration',
                                },
                            },
                            required: ['class'],
                            additionalProperties: false,
                        },
                        {
                            type: 'object',
                            properties: {
                                path: {
                                    type: 'string',
                                },
                                types: {
                                    $ref: '#/definitions/typeImportConfiguration',
                                },
                            },
                            required: ['path'],
                            additionalProperties: false,
                        },
                    ],
                },
            },
            type: 'array',
            items: {
                anyOf: [
                    {
                        $ref: '#/definitions/moduleConfiguration',
                    },
                    {
                        type: 'array',
                        items: {
                            $ref: '#/definitions/moduleConfiguration',
                        },
                    },
                ],
            },
        },
    },
    create(context) {
        const groupConfigurations = context.options.length ? context.options : defaultConfiguration;
        const source = context.getSourceCode();
        const imports = (0, ast_1.importModules)(source).map((node) => {
            return {
                index: groupIndex(node, groupConfigurations),
                node,
            };
        });
        if (imports.length === 0) {
            return {};
        }
        const sorted = (0, sort_1.sortByPath)(imports, ['index']);
        let previousIndex = sorted[0].index;
        const groups = sorted.reduce((r, v) => {
            const current = r[r.length - 1];
            if (previousIndex === v.index) {
                current.push(v.node);
            }
            else {
                r.push([v.node]);
            }
            previousIndex = v.index;
            return r;
        }, [[]]);
        if (sorted.some((node, i) => node !== imports[i])) {
            (0, rule_1.fixRange)(context, {
                range: (0, ast_1.extrema)(imports.map(({ node }) => node)),
                message: `Expected import groups: ${groupLabels(groupConfigurations).join(', ')}`,
                code: groups.map((nodes) => nodes.map((node) => source.getText(node)).join('\n')).join('\n\n'),
            });
        }
        else {
            groups.forEach((group, i) => {
                for (let j = 1; j < group.length; j++) {
                    const previous = group[j - 1];
                    const current = group[j];
                    checkLines(context, previous, current, 0);
                }
                if (i === 0) {
                    return;
                }
                const previous = groups[i - 1][groups[i - 1].length - 1];
                const [current] = group;
                checkLines(context, previous, current, 1);
            });
        }
        return {};
    },
};
