// eslint-plugin-style/dist/rules/sort-imports.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rule = exports.TypeImportInlinePosition = exports.TypeImportGroupPosition = void 0;
const ast_1 = require("../util/ast");
const rule_1 = require("../util/rule");
const sort_1 = require("../util/sort");
var TypeImportGroupPosition;
(function (TypeImportGroupPosition) {
    TypeImportGroupPosition["Ignore"] = "ignore";
    TypeImportGroupPosition["Top"] = "top";
    TypeImportGroupPosition["Bottom"] = "bottom";
    TypeImportGroupPosition["AboveValue"] = "above-value";
    TypeImportGroupPosition["BelowValue"] = "below-value";
})(TypeImportGroupPosition = exports.TypeImportGroupPosition || (exports.TypeImportGroupPosition = {}));
var TypeImportInlinePosition;
(function (TypeImportInlinePosition) {
    TypeImportInlinePosition["Ignore"] = "ignore";
    TypeImportInlinePosition["Start"] = "start";
    TypeImportInlinePosition["End"] = "End";
})(TypeImportInlinePosition = exports.TypeImportInlinePosition || (exports.TypeImportInlinePosition = {}));
const defaultConfiguration = {
    specifier: 'source',
    locales: ['en-US'],
    numeric: true,
    caseFirst: 'lower',
    sortExports: true,
    typesInGroup: TypeImportGroupPosition.Ignore,
    inlineTypes: TypeImportInlinePosition.Ignore,
};
exports.rule = {
    meta: {
        fixable: 'code',
        schema: [
            {
                type: 'object',
                properties: {
                    locales: {
                        type: 'array',
                        items: {
                            type: 'string',
                        },
                    },
                    sensitivity: {
                        enum: ['base', 'accent', 'case', 'variant'],
                    },
                    ignorePunctuation: {
                        type: 'boolean',
                    },
                    numeric: {
                        type: 'boolean',
                    },
                    caseFirst: {
                        enum: ['upper', 'lower', 'false'],
                    },
                    caseGroups: {
                        type: 'boolean',
                    },
                    specifier: {
                        enum: ['source', 'rename'],
                    },
                    sortExports: {
                        type: 'boolean',
                    },
                    typesInGroup: {
                        enum: ['ignore', 'top', 'bottom', 'above-value', 'below-value'],
                    },
                },
            },
        ],
    },
    create(context) {
        const configuration = Object.assign(Object.assign({}, defaultConfiguration), context.options[0]);
        const source = context.getSourceCode();
        const partition = (result, node, index, from) => {
            if (index > 0 && (0, ast_1.linesBetween)(from[index - 1], node) > 0) {
                result.push([]);
            }
            result[result.length - 1].push(node);
            return result;
        };
        const sortModules = (group) => {
            const sorted = (0, sort_1.sortByPath)(group, ['source', 'value'], configuration);
            if (configuration.typesInGroup !== TypeImportGroupPosition.Ignore) {
                sorted.sort((a, b) => {
                    const aIsType = (0, ast_1.isTypeImportOrExport)(a);
                    const bIsType = (0, ast_1.isTypeImportOrExport)(b);
                    if (aIsType === bIsType) {
                        return 0;
                    }
                    // eslint-disable-next-line @typescript-eslint/switch-exhaustiveness-check
                    switch (configuration.typesInGroup) {
                        case TypeImportGroupPosition.Top:
                            return aIsType ? -1 : 1;
                        case TypeImportGroupPosition.Bottom:
                            return aIsType ? 1 : -1;
                    }
                    if (a.source.value === b.source.value) {
                        // eslint-disable-next-line @typescript-eslint/switch-exhaustiveness-check
                        switch (configuration.typesInGroup) {
                            case TypeImportGroupPosition.AboveValue:
                                return aIsType ? -1 : 1;
                            case TypeImportGroupPosition.BelowValue:
                                return aIsType ? 1 : -1;
                        }
                    }
                    return 0;
                });
            }
            if (sorted.some((node, i) => node !== group[i])) {
                (0, rule_1.fixRange)(context, {
                    range: (0, ast_1.extrema)(group),
                    message: 'Expected modules in group to be sorted',
                    code: sorted.map((node) => source.getText(node)).join('\n'),
                });
            }
        };
        const sortSpecifiers = (specifiers) => {
            if (!specifiers.length) {
                return;
            }
            const sorted = (() => {
                if (specifiers[0].type === 'ImportSpecifier') {
                    const from = configuration.specifier === 'source' ? 'imported' : 'local';
                    return (0, sort_1.sortByPath)(specifiers, [from, 'name'], configuration);
                }
                const from = configuration.specifier === 'source' ? 'local' : 'exported';
                return (0, sort_1.sortByPath)(specifiers, [from, 'name'], configuration);
            })();
            if (configuration.inlineTypes !== TypeImportInlinePosition.Ignore) {
                sorted.sort((a, b) => {
                    const aIsType = (0, ast_1.isTypeImportOrExport)(a);
                    const bIsType = (0, ast_1.isTypeImportOrExport)(b);
                    if (aIsType === bIsType) {
                        return 0;
                    }
                    // eslint-disable-next-line @typescript-eslint/switch-exhaustiveness-check
                    switch (configuration.inlineTypes) {
                        case TypeImportInlinePosition.Start:
                            return aIsType ? -1 : 1;
                        case TypeImportInlinePosition.End:
                            return aIsType ? 1 : -1;
                    }
                    return 0;
                });
            }
            if (sorted.some((node, i) => node !== specifiers[i])) {
                (0, rule_1.fixRange)(context, {
                    range: (0, ast_1.extrema)(specifiers),
                    message: 'Expected specifiers to be sorted',
                    code: sorted.map((node) => source.getText(node)).join(', '),
                });
            }
        };
        (0, ast_1.importModules)(source)
            .reduce(partition, [[]])
            .forEach((group) => {
            sortModules(group);
            group.forEach((node) => {
                sortSpecifiers(node.specifiers.filter(($) => $.type === 'ImportSpecifier'));
            });
        });
        if (configuration.sortExports) {
            const sortExportSpecifiers = (specifiers) => {
                const from = configuration.specifier === 'source' ? 'local' : 'exported';
                const sorted = (0, sort_1.sortByPath)(specifiers, [from, 'name'], configuration);
                if (sorted.some((node, i) => node !== specifiers[i])) {
                    (0, rule_1.fixRange)(context, {
                        range: (0, ast_1.extrema)(specifiers),
                        message: 'Expected specifiers to be sorted',
                        code: sorted.map((node) => source.getText(node)).join(', '),
                    });
                }
            };
            (0, ast_1.exportModules)(source)
                .reduce(partition, [[]])
                .forEach((group) => {
                sortModules(group);
                group.forEach((node) => { var _a; return sortExportSpecifiers((_a = node.specifiers) !== null && _a !== void 0 ? _a : []); });
            });
        }
        return {};
    },
};
